src/sfizz/modulations/ModMatrix.cpp

```diff
@@ -0,0 +1,427 @@
+            }
+        }
+    }
+
+    impl.numFrames_ = 0;
+}
+
+void ModMatrix::beginVoice(NumericId<Voice> voiceId, NumericId<Region> regionId)
+{
+    Impl& impl = *impl_;
+
+    impl.voiceId_ = voiceId;
+    impl.regionId_ = regionId;
+
+    for (Impl::Source &source : impl.sources_) {
+        const int flags = source.key.flags();
+        if (flags & kModIsPerVoice)
+            source.bufferReady = false;
+    }
+    for (Impl::Target &target : impl.targets_) {
+        const int flags = target.key.flags();
+        if (flags & kModIsPerVoice)
+            target.bufferReady = false;
+    }
+}
+
+void ModMatrix::endVoice()
+{
+    Impl& impl = *impl_;
+    const uint32_t numFrames = impl.numFrames_;
+    const NumericId<Voice> voiceId = impl.voiceId_;
+    const NumericId<Region> regionId = impl.regionId_;
+
+    for (Impl::Source &source : impl.sources_) {
+        if (!source.bufferReady) {
+            const int flags = source.key.flags();
+            if ((flags & kModIsPerVoice) && source.key.region() == regionId) {
+                absl::Span<float> buffer(source.buffer.data(), numFrames);
+                source.gen->generateDiscarded(source.key, voiceId, buffer);
+            }
+        }
+    }
+
+    impl.voiceId_ = {};
+    impl.regionId_ = {};
+}
+
+float* ModMatrix::getModulation(TargetId targetId)
+{
+    if (!validTarget(targetId))
+        return nullptr;
+
+    Impl& impl = *impl_;
+    const NumericId<Region> regionId = impl.regionId_;
+    const uint32_t targetIndex = targetId.number();
+    Impl::Target &target = impl.targets_[targetIndex];
+    const int targetFlags = target.key.flags();
+
+    const uint32_t numFrames = impl.numFrames_;
+    absl::Span<float> buffer(target.buffer.data(), numFrames);
+
+    // only accept per-voice targets of the same region
+    if ((targetFlags & kModIsPerVoice) && regionId != target.key.region())
+        return nullptr;
+
+    // check if already processed
+    if (target.bufferReady)
+        return buffer.data();
+
+    // set the ready flag to prevent a cycle
+    // in case there is, be sure to initialize the buffer
+    target.bufferReady = true;
+
+    auto sourcesPos = target.connectedSources.begin();
+    auto sourcesEnd = target.connectedSources.end();
+    bool isFirstSource = true;
+
+    // generate first source in output buffer, next sources in temporary buffer
+    // then add or multiply, depending on target flags
+    while (sourcesPos != sourcesEnd) {
+        Impl::Source &source = impl.sources_[sourcesPos->first];
+        const float sourceDepth = sourcesPos->second.sourceDepth_;
+        const int sourceFlags = source.key.flags();
+
+        // only accept per-voice sources of the same region
+        bool useThisSource = true;
+        if (sourceFlags & kModIsPerVoice)
+            useThisSource = (regionId == source.key.region());
+
+        if (useThisSource) {
+            if (isFirstSource) {
```

Comment:
Could shortcut here if `depth == 0.0f`

Reply:
Yes not only you're correct, but there is a greater problem.
The source is supposed to be buffered per region or cycle, not have its generation output used directly.
This makes the source generate more than once if in use by multiple targets.

ie. make use of Source `buffer`/`bufferReady`

Reply:
See a97f506.
Regarding the shortcut: a future consideration is that depth can be modulated.
When this is implemented, `sourceDepth`=0 is an insufficient condition, one has also to check that depth is free of any modulations.

I'm sure I to have this idea in mind previously, but postponed it until making the depth modulation.
There is not an immediate benefit anyway, the source will have to implement a fast path in `generateDiscarded`.

As CC as concerned, I believe the fast path is implementable in case of not smoothing, or smoothing near completed.